### DISCUSSION TITLE
Dynamic OData

### DISCUSSION THREAD

**Discussion by fabiocbr75** [OData/AspNetCoreOData#637] [Q&A]
Hello,
I need to create an OData endpoint dynamically.
I see that there are two possibilities:

- https://github.com/microsoft/dynamic-odata
- https://github.com/OData/AspNetCoreOData/tree/main/sample/ODataDynamicModel

I don't understand what are the differences between the two projects and what is the best way to perform my job.

Can you help me to understand the difference?

Regards

**Comment by vigouredelaruse** [comment]
for my own future reference as well

the first link referred to by the poster references a repo with no recent commits that
* purports to support dynamic odata endpoints with a client library that renders odata metedata based on runtime configuration (presumably as opposed to statically defined at compile time)
* the second link describes how to expose an odata compliant endpoint (ODataController) for an arbitrary entity set **and datasource** such that the controller responds to requests shaped like // Get entityset(key) odata/{datasource}/{entityset}({key})
* a request such as // Get entityset(key) odata/{datasource}/{entityset}({key}) does not require that the responding odatacontroller be provided with an edm model containing definitions for every member of the entityset it serves. that's the kind of digital gun that pays for itself on the very first request

the second link referenced by the original poster points to a repo with the below area of interest
https://github.com/OData/AspNetCoreOData/blob/main/sample/ODataDynamicModel/Controllers/HandleAllController.cs

wherein there is useful code pertaining to the use of the asp.net core HttpContext OData Feature that exposes registered EdmModels (for a given ODataPath) and permits reflection style code over the same as per
```
          ODataPath path = Request.ODataFeature().Path;
            IEdmCollectionType collectionType = (IEdmCollectionType)path.Last().EdmType;
            IEdmEntityTypeReference edmEntityTypeReference = collectionType.ElementType.AsEntity();
            var edmEntityType = edmEntityTypeReference.EntityDefinition();

            //Set the SelectExpandClause on OdataFeature to include navigation property set in the $expand
            SetSelectExpandClauseOnODataFeature(path, edmEntityType);

```

**Comment by lorenzocas** [comment]
Hello,
i'm trying to adapt https://github.com/OData/AspNetCoreOData/tree/main/sample/ODataDynamicModel to use entity framework, but i'm not sure how to implement the ef querying part in the sample. I think i should change the Get() Methods in MyDataSource class, but i'm not sure how. Could you please give me some hints how to do this ?

Thanks
Best regards

**Reply by xuzhg** [reply]
@lorenzocas can you file a new issue to share us more details about what you want to achieve?

**Comment by lorenzocas** [comment]
Hello,
I created a new discussion here:

https://github.com/OData/AspNetCoreOData/discussions/1317

Thanks
